pkg/frontend/admin_openshiftcluster_resources_list_test.go

```diff
@@ -182,19 +182,19 @@ func TestAdminListResourcesList(t *testing.T) {
 			vnetClientFactory := mocknetwork.NewMockVirtualNetworksClient(controller)
 			tt.mocks(tt, openshiftClusters, subscriptions, resourcesClient, computeClient, vnetClientFactory)
 
-			f, err := NewFrontend(ctx, logrus.NewEntry(logrus.StandardLogger()), env, &database.Database{
+			f, err := New(ctx, logrus.NewEntry(logrus.StandardLogger()), env, &database.Database{
 				OpenShiftClusters: openshiftClusters,
 				Subscriptions:     subscriptions,
-			}, api.APIs, &noop.Noop{}, nil, nil,
-				func(subscriptionID string, authorizer autorest.Authorizer) features.ResourcesClient {
+			}, api.APIs, &noop.Noop{}, nil,
+				WithResourcesClientFactory(func(subscriptionID string, authorizer autorest.Authorizer) features.ResourcesClient {
 					return resourcesClient
-				},
-				func(subscriptionID string, authorizer autorest.Authorizer) compute.VirtualMachinesClient {
+				}),
+				WithVMClientFactory(func(subscriptionID string, authorizer autorest.Authorizer) compute.VirtualMachinesClient {
 					return computeClient
-				},
-				func(subscriptionID string, authorizer autorest.Authorizer) network.VirtualNetworksClient {
+				}),
+				WithVNetClientFactory(func(subscriptionID string, authorizer autorest.Authorizer) network.VirtualNetworksClient {
 					return vnetClientFactory
-				},
+				}),
```

Comment:
I think I prefer the chaining idea that @hawkowl pointed out..

```go
f, err := New(ctx, ..)
if err != nil {
    return err
}
f.WithResourcesClientFactory(resourcesClient).
    WithVMClientFactory(computeClient).
    WithVNetClientFactory(vnetClientFactory)

Reply:
I like this idea.  If we ever need to add anything we could just create another function that'll chain to the `New` function instead of trying to figure out what "Options" are necessary for certain things appended to the signature of the constructor.

Reply:
+1

Reply:
I will try this out - it does look better.

@hawkowl I am confused by this, though:

> And then, when you call f.Run(ctx, ...) it checks to see what clients it has not got set, and then generates the default version.

Why would `Run` need to check or generate anything? `New` could still set the default, "real" factory functions. Then those could be overridden by the chaining functions (in unit tests). The factory functions aren't called (that is, they don't create any clients) until an API request is handled. There's no harm in setting the value of those fields twice for unit tests (first the default, then the override). Getting `Run` involved is confusing.